<a href="https://colab.research.google.com/github/oleg-zao/My-Python-Home-Work/blob/main/hw_3_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

In [7]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [8]:
import pandas as pd

# Зчитуємо дані у змінну df типу pandas.DataFrame
df = pd.read_csv('/content/drive/MyDrive/supermarket_sales.csv')

# Переглядаємо перші кілька рядків DataFrame
display(df.head())

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [12]:
# Сортуємо дані за 'Unit price' у порядку спадання
sorted_df = df.sort_values(by='Unit price', ascending=False)

# Виводимо перші 5 рядків
top_5 = sorted_df.head(5)

display(top_5)

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [13]:
# Фільтруємо дані для категорії 'Electronic accessories'
electronic_df = df[df['Product line'] == 'Electronic accessories']

# Знаходимо рядок з найбільшою кількістю товару та найбільшою ціною
max_row = electronic_df.loc[(electronic_df['Quantity'] == electronic_df['Quantity'].max()) &
                            (electronic_df['Unit price'] == electronic_df['Unit price'].max())]

# Виводимо цей рядок на екран
display(max_row)


Empty DataFrame
Columns: [Invoice ID, Branch, City, Customer type, Gender, Product line, Unit price, Quantity, Tax 5%, Total, Date, Time, Payment, cogs, gross margin percentage, gross income, Rating]
Index: []

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [14]:
# Створюємо функцію для категоризації рейтингу
def categorize_rating(rating):
    if rating >= 8:
        return 'High'
    elif 6 <= rating < 8:
        return 'Medium'
    else:
        return 'Low'

# Створюємо нову колонку 'Rating Category' за допомогою функції
df['Rating Category'] = df['Rating'].apply(categorize_rating)

# Переглядаємо перші кілька рядків з новою колонкою
display(df.head())


Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  
0            High  
1            High  
2          Medium  
3            High  
4             Low

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [15]:
# Обчислюємо відсоток кожної категорії рейтингу
rating_category_percent = df['Rating Category'].value_counts(normalize=True) * 100

# Виводимо результат
print(rating_category_percent)


Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [16]:
time = '11:28'

In [17]:
# Функція для витягування годин із часу формату "година:хвилина"
def get_hour(time_str):
    # Розділяємо рядок за двокрапкою і беремо перший елемент (годину)
    hour = time_str.split(':')[0]
    # Конвертуємо його у тип int і повертаємо
    return int(hour)

# Перевіримо роботу функції на змінній time
time = '11:28'
hour = get_hour(time)
print(hour)


11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [20]:
# Спочатку переконаємося, що функція get_hour вже визначена
def get_hour(time_str):
    # Розділяємо рядок за двокрапкою і беремо перший елемент (годину)
    hour = time_str.split(':')[0]
    # Конвертуємо його у тип int і повертаємо
    return int(hour)

# Використовуємо apply() для створення нової колонки 'Hour'
df['Hour'] = df['Time'].apply(get_hour)

# Переглянемо перші кілька рядків, щоб перевірити нову колонку
display(df.head())


Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [22]:
# Рахуємо кількість покупок за кожною годиною
hourly_counts = df['Hour'].value_counts().sort_index()

# Виводимо результат
display(hourly_counts)


Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: count, dtype: int64

З огляду на результати:

- Найбільше покупок було здійснено о **19:00** (113 покупок).
- Інші пікові години: **13:00** (103 покупок), **15:00** (102 покупок), **10:00** (101 покупка).
- Найменше покупок о **17:00** (74 покупки).

### Чому саме ці години мають найбільше покупок:
1. **19:00** — це вечірній час, коли багато людей завершують роботу, повертаються додому або мають час зайти до магазину. Це логічно, адже вечір — зручний час для покупок після робочого дня.
2. **13:00** та **15:00** — це періоди обідньої перерви та післяобіднього часу. Можливо, в ці години люди також виходять на обід або роблять короткі перерви для покупок.
3. **10:00** — це ранковий час, коли багато людей можуть робити покупки після відкриття магазинів або початку робочого дня.

### Логічність піків:
Так, це логічно, що покупки зосереджені в обідній час і вечірні години, оскільки це найзручніші періоди для багатьох покупців. Більшість людей не мають можливості робити покупки під час активної роботи або на початку дня.



5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.